cmd/zfs/zfs_main.c

```diff
@@ -7212,23 +7233,34 @@ share_mount(int op, int argc, char **argv)
 			argv++;
 		}
 
-		if (argc > 1) {
+		if (!recursive && argc > 1) {
 			(void) fprintf(stderr, gettext("usage: "
-			    "zfs mount -a <dataset>\n"));
+			    "zfs mount -a [filesystem]\n"));
 			usage(B_FALSE);
 		}
 
 		if (recursive && argc != 1) {
 			(void) fprintf(stderr, gettext("usage: "
-			    "zfs mount -r <dataset>\n"));
+			    "zfs mount -r <filesystem>\n"));
 			usage(B_FALSE);
 		}
 
-		const char *dataset = (argv[0] != NULL) ? argv[0] : NULL;
+		/*
+		 * Enable recursive -a for mounting only
+		 * Validate filesystem is actually a valid zfs filesystem
+		 */
+		const char *filesystem = NULL;
+		if (op == OP_MOUNT)
+			filesystem = argv[0];
```

Comment:
Is there a reason this is limited only to `mount`?

Reply:
Yes/no, this pull request was originally for mount only, and `zfs share` has additional args `[smb|nfs]` that I didn't see where it was pulled from.
I thought it best to limit the PR to `zfs mount` untill/unless I look into how share works.

Reply:
There are no additional arguments documented for `zfs share`. The control flow is exactly the same except for the part that actually acts on each filesystem, so limiting this to mount-only seems arbitrary and causes an unnecessary split in the nature of the two recursive filesystem operations.

Reply:
Agreed, I didn't see anything in the manpages, but usage differs?
Hence why I disabled it until I can investigate it properly.

Reply:
By the time you're consuming this argument, any `nfs` or `smb` argument has already been consumed zfs `zfs share`, so the syntax for `zfs share` becomes
```sh
zfs share [-l] <filesystem | <-a|-r> [nfs|smb] [filesystem]>
```
assuming, of course, that you don't drop the `-r`.

Reply:
`-r` is not currently enabled for `zfs share`, I was using `op == OP_MOUNT` to restrict `-a [filesystem]`.

I believe this would be the correct syntax for the `-a` _arg_;
````
zfs share [-l] <-a [nfs|smb] [filesystem] | filesystem>
````

My original reasoning for disabling `-a [filesystem]`
- I didn't have time to track down how the args work to verify if my usage is correct
- I would need to update the man pages adding the missing information [nfs|smb]
- I haven't use `zfs share` (something I should do when I have spare time)